### PR TITLE
make `output_filenames` a real query

### PR DESCRIPTION
**Pull request by oli-obk**
part of #105462

This may be a perf regression and is not obviously the right way forward. We may store this information in the resolver after freezing it for example.

### CODE REVIEW

File: compiler/rustc_driver/src/lib.rs [hunk 1]
```diff
@@ -333,6 +332,8 @@ fn run_compiler(
                 return early_exit();
             }
 
+            queries.global_ctxt()?.enter(|tcx| tcx.output_filenames_raw(()))?;
```
Comment (petrochenkov) [inline compiler/rustc_driver/src/lib.rs:335]: Could you add a comment explaining that this is needed for `tcx.output_filenames()` to never fail?

Comment (petrochenkov) [reply compiler/rustc_driver/src/lib.rs:335]: Nit: also `let gctxt = queries.global_ctxt()?;` with `gctxt` reused throughout this function would be nicer than calling `queries.global_ctxt()?` multiple times.

File: compiler/rustc_interface/src/lib.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #![feature(box_patterns)]
 #![feature(decl_macro)]
 #![feature(internal_output_capture)]
+#![feature(lint_reasons)]
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/lib.rs:4]: Where is it used?

File: compiler/rustc_interface/src/passes.rs [hunk 3]
```diff
@@ -663,13 +662,11 @@ fn write_out_deps(
...
-pub fn prepare_outputs(
-    sess: &Session,
-    krate: &ast::Crate,
-    boxed_resolver: &RefCell<BoxedResolver>,
-    crate_name: Symbol,
-) -> Result<OutputFilenames> {
+fn output_filenames(tcx: TyCtxt<'_>, (): ()) -> Result<&Arc<OutputFilenames>> {
+    let sess = tcx.sess;
     let _timer = sess.timer("prepare_outputs");
+    let krate = &*tcx.resolver_for_lowering(()).borrow().1;
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/passes.rs:668]:
```suggestion
    let (_, krate) = &*tcx.resolver_for_lowering(()).borrow();
```
Destructuring almost always looks better than tuple fields.

File: compiler/rustc_interface/src/passes.rs [hunk 4]
```diff
@@ -663,13 +662,11 @@ fn write_out_deps(
     }
 }
 
-pub fn prepare_outputs(
-    sess: &Session,
-    krate: &ast::Crate,
-    boxed_resolver: &RefCell<BoxedResolver>,
-    crate_name: Symbol,
-) -> Result<OutputFilenames> {
+fn output_filenames(tcx: TyCtxt<'_>, (): ()) -> Result<&Arc<OutputFilenames>> {
```
Comment (petrochenkov) [inline compiler/rustc_interface/src/passes.rs:665]:
```suggestion
fn output_filenames(tcx: TyCtxt<'_>, (): ()) -> Result<&OutputFilenames> {
```
Is `Arc` still needed here?
Does it need to be cloned to some place where `tcx` and its arenas are longer available?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 5]
```diff
@@ -2043,6 +2045,10 @@ impl<'tcx> TyCtxt<'tcx> {
         self.typeck(self.hir().body_owner_def_id(body))
     }
 
+    pub fn output_filenames(self) -> &'tcx Arc<OutputFilenames> {
+        self.output_filenames_raw(()).unwrap()
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/mod.rs:2049]: Could you add a comment explaining why this unwrap never fails?

File: compiler/rustc_interface/src/passes.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_interface/src/passes.rs:665]: Yes, llvm, ssa and cranelift all store a clone of it it until after the TyCtxt is gone.

File: compiler/rustc_interface/src/passes.rs (hunk 3 above)
Comment (fbstj) [reply compiler/rustc_interface/src/passes.rs]: do the similar field usage in 7cc499bb2182c0e4f0468d80a1a5ce95d5fa331e count?

File: compiler/rustc_interface/src/passes.rs [hunk 6]
```diff
@@ -681,45 +678,41 @@ pub fn prepare_outputs(
     if let Some(ref input_path) = sess.io.input.opt_path() {
         if sess.opts.will_create_output_file() {
             if output_contains_path(&output_paths, input_path) {
-                let reported = sess.emit_err(InputFileWouldBeOverWritten { path: input_path });
-                return Err(reported);
+                sess.emit_err(InputFileWouldBeOverWritten { path: input_path });
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_interface/src/passes.rs]: I switched to fatal errors and added tests for all but the `--out-dir` code path, as that one would require a run-make test

[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_interface/src/passes.rs]: Nevermind, we need that anyway :)

**Comment by petrochenkov** [comment]
r=me after squashing review commits into main commits.
@rustbot author

File: compiler/rustc_middle/src/ty/mod.rs [hunk 7]
```diff
@@ -2043,6 +2045,10 @@ impl<'tcx> TyCtxt<'tcx> {
         self.typeck(self.hir().body_owner_def_id(body))
     }
 
+    pub fn output_filenames(self) -> &'tcx Arc<OutputFilenames> {
+        self.output_filenames_raw(())
+    }
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/mod.rs:2050]: Nit: As `output_filenames_raw` cannot return an `Err` any more, do we need to keep the `output_filenames_raw`/`output_filenames` distinction?

File: compiler/rustc_interface/src/passes.rs [hunk 8]
```diff
@@ -681,45 +678,41 @@ pub fn prepare_outputs(
...
             if fs::create_dir_all(dir).is_err() {
-                let reported = sess.emit_err(OutDirError);
-                return Err(reported);
+                sess.emit_fatal(OutDirError);
             }
         }
     }
 
-    Ok(outputs)
+    tcx.arena.alloc(outputs.into())
```
Comment (cjgillot) [inline compiler/rustc_interface/src/passes.rs:708]: Nit: use `arena_cache` to define the query, and avoid manual arena allocation.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 7 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/ty/mod.rs:2050]: It's mostly a small annoyance that we have to call the query as `output_filenames(())`

Comment (oli-obk) [reply compiler/rustc_middle/src/ty/mod.rs:2050]: Considering that's what we had before this PR, I'll revert

**Pull request merged**